### PR TITLE
serialize: fix loading of cloud metadata from .dvc/dvc.lock files

### PR DESCRIPTION
**Pull request by skshetry**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by skshetry** [comment]
Investigating a potential bug, where this patch makes it always push to remote and update version id.

File: dvc/output.py [hunk 1]
```diff
@@ -396,13 +398,19 @@ def __init__(  # noqa: C901
...
         if self.files:
             tree = Tree.from_list(self.files, hash_name=self.hash_name)
             tree.digest()
             self.hash_info = tree.hash_info
             self.meta.isdir = True
             self.meta.nfiles = len(self.files)
             self.meta.size = sum(file.get("size") for file in self.files)
+            self.meta.remote = first(
+                first(file.get("cloud", {})) for file in self.files
+            )
```
Comment (skshetry) [inline dvc/output.py:413]: All of this serialization/deserialization logic should be outside. Will try to look into it after this PR. It's not very important, but it feels like we are struggling with this, as we have to maintain `.dvc` and `dvc.lock` files.

File: dvc/output.py [hunk 2]
```diff
@@ -398,13 +399,19 @@ def __init__(  # noqa: C901
...
+
+    def _compute_meta_hash_info_from_files(self) -> None:
         if self.files:
             tree = Tree.from_list(self.files, hash_name=self.hash_name)
             tree.digest()
             self.hash_info = tree.hash_info
             self.meta.isdir = True
             self.meta.nfiles = len(self.files)
             self.meta.size = sum(file.get("size") for file in self.files)
+            self.meta.remote = first(
```
Comment (skshetry) [inline dvc/output.py:412]: We don't care about `Tree.meta.remote`, but just adding it to symmetry.

**Review by pmrowla — APPROVED** [review]

File: dvc/stage/loader.py [hunk 3]
```diff
@@ -74,12 +77,15 @@ def fill_from_lock(stage, lock_data=None):
...
             hash_value = getattr(item.meta, item.hash_name, None)
             item.hash_info = HashInfo(item.hash_name, hash_value)
-            if item.hash_info and item.hash_info.isdir:
-                item.meta.isdir = True
-            item.files = get_in(checksums, [key, path, item.PARAM_FILES])
+            files = get_in(checksums, [key, path, item.PARAM_FILES], None)
+            if files:
+                item.files = [merge_file_meta_from_cloud(f) for f in files]
+            # pylint: disable-next=protected-access
+            item._compute_meta_hash_info_from_files()
```
Comment (daavoo) [inline dvc/stage/loader.py:88]: minor but the name is kind of misleading here

Comment (skshetry) [reply dvc/stage/loader.py:88]: yeah, I could not think of a good naming. Any suggestions? 🙂

**Pull request merged**